## Resume Screening application

In [1]:
pip install --upgrade numpy pandas

     ---------------------------------------- 0.0/60.8 kB ? eta -:--:--
     ------------ ------------------------- 20.5/60.8 kB 682.7 kB/s eta 0:00:01
     -------------------------------------- 60.8/60.8 kB 816.2 kB/s eta 0:00:00
   ---------------------------------------- 0.0/12.9 MB ? eta -:--:--
    --------------------------------------- 0.3/12.9 MB 5.9 MB/s eta 0:00:03
   -- ------------------------------------- 0.8/12.9 MB 8.1 MB/s eta 0:00:02
   ---- ----------------------------------- 1.3/12.9 MB 9.5 MB/s eta 0:00:02
   ------ --------------------------------- 2.0/12.9 MB 10.6 MB/s eta 0:00:02
   -------- ------------------------------- 2.6/12.9 MB 11.0 MB/s eta 0:00:01
   --------- ------------------------------ 3.2/12.9 MB 11.3 MB/s eta 0:00:01
   ----------- ---------------------------- 3.8/12.9 MB 11.5 MB/s eta 0:00:01
   ------------- -------------------------- 4.4/12.9 MB 11.7 MB/s eta 0:00:01
   --------------- ------------------------ 4.9/12.9 MB 11.6 MB/s eta 0:00:01

In [1]:
pip install scikit-learn


     ---------------------------------------- 0.0/60.8 kB ? eta -:--:--
     -------------------- ------------------- 30.7/60.8 kB 1.4 MB/s eta 0:00:01
     ---------------------------------------- 60.8/60.8 kB 1.6 MB/s eta 0:00:00
  Using cached joblib-1.4.2-py3-none-any.whl.metadata (5.4 kB)
  Using cached threadpoolctl-3.5.0-py3-none-any.whl.metadata (13 kB)
   ---------------------------------------- 0.0/11.1 MB ? eta -:--:--
   - -------------------------------------- 0.3/11.1 MB 6.5 MB/s eta 0:00:02
   -- ------------------------------------- 0.8/11.1 MB 8.7 MB/s eta 0:00:02
   ----- ---------------------------------- 1.4/11.1 MB 10.2 MB/s eta 0:00:01
   ------- -------------------------------- 2.0/11.1 MB 10.8 MB/s eta 0:00:01
   --------- ------------------------------ 2.7/11.1 MB 11.4 MB/s eta 0:00:01
   ----------- ---------------------------- 3.2/11.1 MB 12.0 MB/s eta 0:00:01
   ------------ --------------------------- 3.6/11.1 MB 11.6 MB/s eta 0:00:01
   -------------- ---

In [6]:
pip install "numpy<2"


     ---------------------------------------- 0.0/61.0 kB ? eta -:--:--
     ------------ ------------------------- 20.5/61.0 kB 320.0 kB/s eta 0:00:01
     -------------------------------------- 61.0/61.0 kB 806.3 kB/s eta 0:00:00
   ---------------------------------------- 0.0/15.8 MB ? eta -:--:--
   ---------------------------------------- 0.2/15.8 MB 5.6 MB/s eta 0:00:03
   - -------------------------------------- 0.7/15.8 MB 8.7 MB/s eta 0:00:02
   --- ------------------------------------ 1.3/15.8 MB 10.5 MB/s eta 0:00:02
   ---- ----------------------------------- 2.0/15.8 MB 11.3 MB/s eta 0:00:02
   ------ --------------------------------- 2.5/15.8 MB 11.5 MB/s eta 0:00:02
   ------- -------------------------------- 3.0/15.8 MB 11.4 MB/s eta 0:00:02
   --------- ------------------------------ 3.7/15.8 MB 11.7 MB/s eta 0:00:02
   ---------- ----------------------------- 4.3/15.8 MB 11.9 MB/s eta 0:00:01
   ------------ --------------------------- 4.8/15.8 MB 11.8 MB/s eta 0:00:0

  You can safely remove it manually.
  You can safely remove it manually.


In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import re

In [4]:
data=pd.read_csv("UpdatedResumeDataSet.csv")
data

Category                                             Resume
0    Data Science  Skills * Programming Languages: Python (pandas...
1    Data Science  Education Details \r\nMay 2013 to May 2017 B.E...
2    Data Science  Areas of Interest Deep Learning, Control Syste...
3    Data Science  Skills • R • Python • SAP HANA • Table...
4    Data Science  Education Details \r\n MCA   YMCAUST,  Faridab...
..            ...                                                ...
957       Testing  Computer Skills: • Proficient in MS office (...
958       Testing  ❖ Willingness to accept the challenges. ❖ ...
959       Testing  PERSONAL SKILLS • Quick learner, • Eagerne...
960       Testing  COMPUTER SKILLS & SOFTWARE KNOWLEDGE MS-Power ...
961       Testing  Skill Set OS Windows XP/7/8/8.1/10 Database MY...

[962 rows x 2 columns]

## Exploring Categories

In [5]:
data["Category"].value_counts()

Category
Java Developer               84
Testing                      70
DevOps Engineer              55
Python Developer             48
Web Designing                45
HR                           44
Hadoop                       42
Blockchain                   40
ETL Developer                40
Operations Manager           40
Data Science                 40
Sales                        40
Mechanical Engineer          40
Arts                         36
Database                     33
Electrical Engineering       30
Health and fitness           30
PMO                          30
Business Analyst             28
DotNet Developer             28
Automation Testing           26
Network Security Engineer    25
SAP Developer                24
Civil Engineer               24
Advocate                     20
Name: count, dtype: int64

In [6]:
data.Category.unique()

array(['Data Science', 'HR', 'Advocate', 'Arts', 'Web Designing',
       'Mechanical Engineer', 'Sales', 'Health and fitness',
       'Civil Engineer', 'Java Developer', 'Business Analyst',
       'SAP Developer', 'Automation Testing', 'Electrical Engineering',
       'Operations Manager', 'Python Developer', 'DevOps Engineer',
       'Network Security Engineer', 'PMO', 'Database', 'Hadoop',
       'ETL Developer', 'DotNet Developer', 'Blockchain', 'Testing'],
      dtype=object)

## Clean the data

In [7]:
def cleanText(txt):
    cleanText= re.sub("http\S+\s"," ",txt)
    cleanText= re.sub("RT|CC"," ",cleanText)
    cleanText= re.sub("#\S+\s"," ",cleanText)
    cleanText= re.sub("@\S+"," ",cleanText)
    cleanText= re.sub('[%s]' % re.escape("""!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~""")," ",cleanText)
    cleanText= re.sub("r'[^\x00-\x7f]"," ",cleanText)
    cleanText= re.sub(r"\s+"," ",cleanText)
    cleanText = re.sub(r"^[\-\•\*\d+.\)]\s*", " ", cleanText, flags=re.M)
    return cleanText
    

In [8]:
data['Resume']= data['Resume'].apply(lambda x: cleanText(x))

### Example resume after cleaning

In [9]:
data['Resume'][0]

'Skills Programming Languages Python pandas numpy scipy scikit learn matplotlib Sql Java JavaScript JQuery Machine learning Regression SVM Naïve Bayes KNN Random Forest Decision Trees Boosting techniques Cluster Analysis Word Embedding Sentiment Analysis Natural Language processing Dimensionality reduction Topic Modelling LDA NMF PCA Neural Nets Database Visualizations Mysql SqlServer Cassandra Hbase ElasticSearch D3 js DC js Plotly kibana matplotlib ggplot Tableau Others Regular Expression HTML CSS Angular 6 Logstash Kafka Python Flask Git Docker computer vision Open CV and understanding of Deep learning Education Details Data Science Assurance Associate Data Science Assurance Associate Ernst Young LLP Skill Details JAVASCRIPT Exprience 24 months jQuery Exprience 24 months Python Exprience 24 monthsCompany Details company Ernst Young LLP description Fraud Investigations and Dispute Services Assurance TECHNOLOGY ASSISTED REVIEW TAR Technology Assisted Review assists in accelerating th

## Words into Categorical values

### Using scikit learn library to solve this supervised -learning problem

In [10]:
from sklearn.preprocessing import LabelEncoder

le= LabelEncoder()

### Encoding Labeles from string into numbers 

In [11]:
le.fit(data['Category'])
data['Category']= le.transform(data["Category"])

In [12]:
data.Category.unique()

array([ 6, 12,  0,  1, 24, 16, 22, 14,  5, 15,  4, 21,  2, 11, 18, 20,  8,
       17, 19,  7, 13, 10,  9,  3, 23])

In [13]:
#'Data Science', 'HR', 'Advocate', 'Arts', 'Web Designing',
 #      'Mechanical Engineer', 'Sales', 'Health and fitness',
  #     'Civil Engineer', 'Java Developer', 'Business Analyst',
   #    'SAP Developer', 'Automation Testing', 'Electrical Engineering',
    #   'Operations Manager', 'Python Developer', 'DevOps Engineer',
    #   'Network Security Engineer', 'PMO', 'Database', 'Hadoop',
    #   'ETL Developer', 'DotNet Developer', 'Blockchain', 'Testing'

## Vectorization

### Generete TFIDF score for Resume

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf=TfidfVectorizer(stop_words='english')
tfidf.fit(data['Resume'])
## Save the scores for each word in requiredText
requiredText= tfidf.transform(data['Resume'])

In [15]:
requiredText.shape ## Scores for each resume 

(962, 7370)

## Split the Dataset into Train and test in order to train our model

In [16]:
from sklearn.model_selection import train_test_split

### train_test_split
#### Split arrays or matrices into random train and test subsets.
    Parameters:
*arrays
sequence of indexables with same length / shape[0]
Allowed inputs are lists, numpy arrays, scipy-sparse matrices or pandas dataframes### .

test_size
float or int, default=None
If float, should be between 0.0 and 1.0 and represent the proportion of the dataset to include in the test split. If int, represents the absolute number of test samples. If None, the value is set to the complement of the train size. If train_size is also None, it will be set to ### 0.25.

train_size
float or int, default=None
If float, should be between 0.0 and 1.0 and represent the proportion of the dataset to include in the train split. If int, represents the absolute number of train samples. If None, the value is automatically set to the complement of the t### est size.

random_state
int, RandomState instance or None, default=None
Controls the shuffling applied to the data before applying the split. Pass an int for reproducible output across multiple function calls. ### See Glossary.

shuffle
bool, default=True
Whether or not to shuffle the data before splitting. If shuffle=False then strat### ify must be None.

stratify
ar
If not None, data is split in a stratified fashion, using this as the class labels.
#### Returns:
splittinglist, length=2 * len(arrays)ray-like, default=None

In [17]:
X_train, X_test, y_train, y_test = train_test_split(requiredText, data['Category'], test_size=0.20, random_state=42)

In [18]:
X_train.shape

(769, 7370)

In [19]:
X_test.shape

(193, 7370)

## Now train the model

In [20]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.multiclass import OneVsRestClassifier
from sklearn.metrics import accuracy_score

clf=OneVsRestClassifier(KNeighborsClassifier())
clf.fit(X_train,y_train)
prediction_on_test=clf.predict(X_test)
print(prediction_on_test)

## Now lets print the accuracy of our model on our test dataset
print(accuracy_score(y_test,prediction_on_test)*100,"%")

[15 15 15 13 14 17 16  2  0 14 13 12 16 23 20  5  6  4 10  9 19  1 10 23
 23 21 22 22  2 12 18  1  8 24 11 23  7 12 24  8 18  6  8 19 24 23 21  1
 15  4 15 22 11  5 15 13  1 19  5 12 22 22 20 24 21 18 12 10 10 20 10  8
  9 21 17 21  0 17 16 14 15 11 11  8 20  3 19  8  0  2  9 10  2 23 20 20
 23 12 18 12  7 16  8 14 18  3 14 19 14 14 15 18  8  2 21 18 23 10 23  5
 11 15 12  3  5  3  7 12 19  8 20 19  3 15  9 19  1 23 21  5 20 15 16  7
  7  8 15 18  1 15 13 20  7  4 18 11  5 15  5 12  9 22 18 21  8 23  4 12
 24 16 15 22  8 22  3 16 23 23 12  7 16 18  5  3 18  8 23 23 20 21  6  7
 23]
98.44559585492227 %


#### Our model is 98.45% accurate in its predictions

## Prediction System

In [21]:
my_resume="""Profile
I am a Computer Science undergraduate at Purdue University with strong programming skills in Java and Python. Experienced in working with MySQL databases and front-end technologies like HTML, CSS, and JavaScript. Actively exploring Cloud computing and Machine Learning. Proficient in Python, REST API development, Spring Boot applications, and react for web development, with firsthand experience in data analysis.
Education
Purdue University Aug 2022- May 2026
Bachelor of Science, Computer Science
GPA: 3.95/4.0
Proud Member of UPE (Upsilon Pi Epsilon)
Relevant Courses: Data Structures and algorithm(honors), Computer Architecture, Intro to C, Computer Networks, Software Engineering
Skills
•
Java
•
Data Structures
•
MySQL
•
Git and GitHub
•
Node.js
•
Python Pandas
•
JavaScript
•
Data Analytics
•
Agile Methodologies
•
C language
•
React.js
•
Spring Boot.
•
Computer Security
Experience
Undergraduate Research Assistant – Purdue University July 2024- Present
•
Contributing to a research project examining how data visualization techniques influence decision-making processes across various fields.
•
Collaborate closely with faculty, analyzing diverse datasets to identify effective visualization methods for non-expert audiences.
•
Evaluate and compare distinct graphs, developing guidelines to optimize clarity, accuracy, and impact in data presentations.
CS Student Scholar Mentor -Purdue University Jan 2024- Present
•
Guiding students in introductory Computer Science courses and Data Structures.
•
Offering support and expertise in Math calculus to enhance mathematical understanding.
•
Providing general academic support, including study strategies and time management.
Undergraduate Teaching Assistant- Purdue University May 2024- Present
•
Assist in managing administrative tasks and supporting students in the year-long Senior Capstone course.
•
Collaborate in designing a structured syllabus and creating Python-based assignments for the Intro to Data Analytics course.
•
Support student learning by providing guidance and mentorship, helping them develop essential data analysis skills.
National Cyber League Spring 2024 – Purdue University April 2024
•
Ranked 229th out of 4,199 teams, placing in the top 95th percentile nationally with a score of 1805/3000 points.
•
Demonstrated technical proficiency in Cryptography, Forensics, Network Traffic Analysis, and Open-Source Intelligence during the competition.
•
Displayed strong teamwork and critical thinking skills in addressing complex cybersecurity challenges.
Projects
SMC Gear Checkout and Studio Booking System Enhancement
•
Enhanced Online Booking Portal: Redesigned the Purdue Sweetwater Music Center’s online portal, implementing a gear checkout system with tier-based access control and real-time tracking, integrated with studio bookings for seamless scheduling.
•
Tech Stack Implementation: Developed the solution using React.js, Node.js, Airtable, Git for version control, and deployed it on Vercel to ensure efficient and reliable performance.
Web Page Indexing and Query Program
•
Developed a web crawler in C language to index multiple web pages, storing word counts in trie data structures for efficient management.
•
Enable user queries to search and match web pages based on gathered statistics, implementing advanced search functionality.
•
Optimized data structures for accurate word count storage and fast query retrieval, enhancing program performance.
Budget Application
•
A personal expense management tool designed to streamline financial tracking and analysis using Python.
•
Utilized Pandas for efficient data manipulation, allowing users to analyze spending patterns and track expenses effortlessly.
•
Integrated Matplotlib to create dynamic, visually appealing charts and graphs that provide clear insights into users' financial health.
•
Designed an intuitive, user-friendly interface for seamless expense entry, categorization, and trend analysis, enhancing user engagement and financial awareness.
Certifications
•
Udemy: Java Networking Issued September 2024
•
Udemy: Data Analysis with Python and Pandas- Certificate Issued March 2024
•
Udemy: Understanding HTML and CSS-Certificate Issued October 2023.
•
Udemy: Core Python for Everyone- Certificate Issued September 2023.
Other Achievements
•
Dean's & Semester Honor's List- Purdue University December 2022-May 2024.
•
Ripple Award March 2024
"""

In [22]:
import pickle
pickle.dump(tfidf, open('tfidf.pkl','wb'))
pickle.dump(clf, open('clf.pkl','wb'))

## Main Application

In [88]:
import pickle

# Load the trained classifier
clf=pickle.load(open('clf.pkl','rb'))

# Load the TFID vectorizer
tfidf= pickle.load(open('tfidf.pkl','rb'))

# clean the resume
cleaned_resume= cleanText(my_resume)
#print(cleaned_resume)

# Transform
input_features=tfidf.transform([cleaned_resume])

# Make prediction
prediction_id=clf.predict(input_features)[0]

# Mapping to category name
category_mapping={
    6:"Data Science",
    12:"HR",
    0:"Advocate",
    1:"Arts",
    24:"Web Designing",
    16:"Mechanical Engineer",
    22:"Sales",
    14:"Health and fitness",
    5:"Civil Engineer",
    15:"Java Developer",
    4:"Business Analyst",
    21:"SAP Developer",
    2:"Automation Testing",
    11:"Electrical Engineering",
    18:"Operations Manager",
    20:"Python Developer",
    8:"DevOps Engineer",
    17:"Network Security Engineer",
    19:"PMO",
    7:"Database",
    13:"Hadoop",
    10:"ETL Developer",
    9:"DotNet Developer",
    3:"Blockchain",
    23:"Testing" 
}
category_name= category_mapping.get(prediction_id,"Unkown")
print("Predicted Category: ",category_name)

Predicted Category:  Data Science
